### PR TITLE
Support basicConstraints in openssl_csr

### PR DESCRIPTION
**Pull request by thomwiggers**
##### SUMMARY
Allows to specify that a certificate is a CA

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

openssl_csr

##### ADDITIONAL INFORMATION

```yaml
- name: Generate CA Certificate CSR
  openssl_csr:
    path: work/vpn/ca/ca.csr
    privatekey_path: work/vpn/ca/ca.key
    privatekey_passphrase: "{{ lookup('passwordstore', 'ansible/vpn/ca_key')}}"
    common_name: "VPN Certificate Authority"
    country_name: "NL"
    organization_name: "Thom Wiggers"
    organizational_unit_name: "Wiggers' Waarachtige Waarheidscertificaten"
    key_usage:
      - keyCertSign
      - cRLSign
    key_usage_critical: yes
    basic_constraints:
      - "CA:TRUE"
```

### CODE REVIEW

File: lib/ansible/modules/crypto/openssl_csr.py [hunk 1]
```diff
@@ -221,6 +226,12 @@
     returned: changed or success
     type: list
     sample: [ 'clientAuth' ]
+basicConstraints:
+    description: Indicates if the certifciate belongs to a CA
```
Comment (Akasurde) [inline lib/ansible/modules/crypto/openssl_csr.py]: s/certifciate/certificate/

File: lib/ansible/modules/crypto/openssl_csr.py [hunk 2]
```diff
@@ -422,6 +442,7 @@ def main():
             keyUsage_critical=dict(aliases=['key_usage_critical'], default=False, type='bool'),
             extendedKeyUsage=dict(aliases=['extKeyUsage', 'extended_key_usage'], type='list'),
             extendedKeyUsage_critical=dict(aliases=['extKeyUsage_critical', 'extended_key_usage_critical'], default=False, type='bool'),
+            basicConstraints=dict(aliases=['basic_constraints'], type='list'),
```
Comment (Akasurde) [inline lib/ansible/modules/crypto/openssl_csr.py:453]: Would it be helpful if we provide user available types of basic constraints ?

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/crypto/openssl_csr.py (hunk 2 above)
Comment (thomwiggers) [reply lib/ansible/modules/crypto/openssl_csr.py:453]: None of the other extension apis do this, it's also not entirely clear to me what constraints are supported.

File: lib/ansible/modules/crypto/openssl_csr.py (hunk 1 above)
Comment (thomwiggers) [reply lib/ansible/modules/crypto/openssl_csr.py]: nicely spotted, thanks

**Comment by thomwiggers** [comment]
ready_for_review

File: lib/ansible/modules/crypto/openssl_csr.py (hunk 2 above)
Comment (Akasurde) [reply lib/ansible/modules/crypto/openssl_csr.py:453]: AFAIK, `critical,CA:TRUE, pathlen:0` can be included in example.

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/crypto/openssl_csr.py (hunk 2 above)
Comment (thomwiggers) [reply lib/ansible/modules/crypto/openssl_csr.py:453]: I checked the [RFC](https://tools.ietf.org/html/rfc5280), and I think while `critical` is technically supported, it's not recommended for basicConstraints like it is for (`extended`)`keyUsage`. `critical` should also be set like `extendedKeyUsage_critical` and `keyUsage_critical`.

I also don't know if `pathlen` is a valid option. The RFC shows `pathLenConstraint`, but the RFC also doesn't necessarily match OpenSSL's behaviour and specifications: for example the RFC shows `cA`, but OpenSSL requires `CA`.

**Comment by MarkusTeufelberger** [comment]
Since the standard allows to set this field to "critical", this should be supported by the module. Also I would prefer to have pathLenConstraint added as a separate option.

"basicConstraints" seems to me to be a bit misleading in the standard anyways, since there is only one single valid constraint in there anyways. Why implement it as a list, if it can only hold 0 or 1 element?

**Comment by thomwiggers** [comment]
I don't think `pathLenConstraint` is a separate option, it's one of the `basicConstraints`.

http://www.alvestrand.no/objectid/2.5.29.19.html

I may expand the example to also include `pathLenConstraint:100` or something like that.

File: lib/ansible/modules/crypto/openssl_csr.py [hunk 3]
```diff
@@ -132,6 +132,18 @@
...
+        required: false
+        aliases: ['basicConstraints']
+        description:
+            - Indicates basic constraints, such as if the certificate is a CA.
+        version_added: 2.5
+    basic_constraints_critical:
+        required: false
+        aliases: [ 'basicConstraints_critical' ]
+        description:
+            - Should the extkeyUsage extension be considered as critical
```
Comment (MarkusTeufelberger) [inline lib/ansible/modules/crypto/openssl_csr.py]: ~~extkeyUsage~~ basic_constraints

File: lib/ansible/modules/crypto/openssl_csr.py [hunk 4]
```diff
@@ -221,6 +233,11 @@
     returned: changed or success
     type: list
     sample: [ 'clientAuth' ]
+basicConstraints:
```
Comment (MarkusTeufelberger) [inline lib/ansible/modules/crypto/openssl_csr.py:236]: this block should be removed

File: lib/ansible/modules/crypto/openssl_csr.py [hunk 5]
```diff
@@ -261,6 +278,8 @@ def __init__(self, module):
         self.keyUsage_critical = module.params['keyUsage_critical']
         self.extendedKeyUsage = module.params['extendedKeyUsage']
         self.extendedKeyUsage_critical = module.params['extendedKeyUsage_critical']
+        self.basicConstraints = module.params['basicConstraints']
+        self.basicConstraints = module.params['basicConstraints_critical']
```
Comment (MarkusTeufelberger) [inline lib/ansible/modules/crypto/openssl_csr.py]: same name as one line above, should be `self.basicConstraints_critical`

File: lib/ansible/modules/crypto/openssl_csr.py (hunk 3 above)
Comment (thomwiggers) [reply lib/ansible/modules/crypto/openssl_csr.py]: oops

File: lib/ansible/modules/crypto/openssl_csr.py (hunk 4 above)
Comment (thomwiggers) [reply lib/ansible/modules/crypto/openssl_csr.py:236]: I've now included it in `dump`

File: lib/ansible/modules/crypto/openssl_csr.py (hunk 5 above)
Comment (thomwiggers) [reply lib/ansible/modules/crypto/openssl_csr.py]: Oops, fixed.

File: lib/ansible/modules/crypto/openssl_csr.py (hunk 4 above)
Comment (MarkusTeufelberger) [reply lib/ansible/modules/crypto/openssl_csr.py:236]: Ah, I was looking in the wrong section, this is the return section, not the documentation one. :-)

**Comment by MarkusTeufelberger** [comment]
Looks fine by me. It would be great if you could also add an integration test for this behaviour/option, but this is not a requirement to merge this code as far as I understand the guidelines for the project.

Still, especially with crypto stuff there are lots of things that can go wrong or unplanned, especially when dealing with different versions of openssl and pyopenssl, sometimes patched downstream too.

From my side I'll give a /shipit, I'd still love to see @Spredzy take a look at this.

**Comment by MarkusTeufelberger** [comment]
bot_status

**Comment by MarkusTeufelberger** [comment]
shipit

**Comment by thomwiggers** [comment]
ready_for_review

**Comment by MarkusTeufelberger** [comment]
There's a merge conflict now, I guess it needs a rebase (probably because of the "file" documentation fragment).

**Comment by thomwiggers** [comment]
I've rebased it.

**Comment by MarkusTeufelberger** [comment]
shipit

**Review by Akasurde — APPROVED** [review]

**Pull request merged**